src/stan/agrad/autodiff.hpp

```diff
@@ -359,6 +357,75 @@ namespace stan {
       stan::agrad::recover_memory_nested();
     }
 
-  }
-}
+    /**
+     * Calculate the value, the Hessian, and the gradient of the Hessian
+     * of the specified function at the specified argument.
+     *
+     * <p>The functor must implement
+     *
+     * <code>
+     * using stan::agrad::fvar;
+     * using stan::agrad::var;
+     *
+     * fvar<fvar<var> >
+     * operator()(const
+     * Eigen::Matrix<fvar<fvar<var> >, Eigen::Dynamic, 1>&)
+     * </code>
+     *
+     * using only operations that are defined for
+     * <code>stan::agrad::fvar</code> and <code>stan::agrad::var</code>.
+     *
+     * This latter constraint usually
+     * requires the functions to be defined in terms of the libraries
+     * defined in Stan or in terms of functions with appropriately
+     * general namespace imports that eventually depend on functions
+     * defined in Stan.
+     *
+     * @tparam F Type of function
+     * @param[in] f Function
+     * @param[in] x Argument to function
+     * @param[out] fx Function applied to argument
+     * @param[out] H Hessian of function at argument
+     * @param[out] grad_H Gradient of the Hessian of function at argument
+     */
+    template <typename F>
+    void
+    grad_hessian(const F& f,
+                 const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
+                 double& fx,
+                 Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>& H,
+                 std::vector<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> >& grad_H) {
+      using Eigen::Matrix;
+      using Eigen::Dynamic;
+     	fx = f(x);
+			int d = x.size();
```

Comment:
I don't know where these come from, my config file for vim should be
automatically exchanging tabs for spaces. Clearly it is failing. Will fix.

On Tuesday, February 10, 2015, Bob Carpenter <notifications@github.com>
wrote:

> In src/stan/agrad/autodiff.hpp
> <https://github.com/stan-dev/stan/pull/1276#discussion_r24472702>:
>
> > +     * @param[in] x Argument to function
> > +     * @param[out] fx Function applied to argument
> > +     * @param[out] H Hessian of function at argument
> > +     * @param[out] grad_H Gradient of the Hessian of function at argument
> > +     */
> > +    template <typename F>
> > +    void
> > +    grad_hessian(const F& f,
> > +                 const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
> > +                 double& fx,
> > +                 Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>& H,
> > +                 std::vector<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> >& grad_H) {
> > +      using Eigen::Matrix;
> > +      using Eigen::Dynamic;
> > +     	fx = f(x);
> > +			int d = x.size();
>
> You've got to lose the tabs or whatever's messing up the code
> indentantation. I don't know if vi can do it automatically. I have emacs
> set up to never save any code files with tabs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1276/files#r24472702>.
>


-- 

Rob